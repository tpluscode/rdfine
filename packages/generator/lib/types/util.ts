import { GraphPointer } from 'clownface'
import { rdf, rdfs, schema } from '@tpluscode/rdf-ns-builders'
import { TypeMap } from '../types/index.js'

export function isDatatype(term: GraphPointer, typeMappings: TypeMap): boolean {
  if (term.has(rdf.type, rdfs.Datatype).values.length) {
    return true
  }

  const result = term.has(rdf.type, [schema.DataType, rdfs.Datatype]).values.length > 0
  if (result) {
    return true
  }

  const superClasses = term.out(rdfs.subClassOf).toArray()
  for (const superClass of superClasses) {
    if (isDatatype(superClass, typeMappings)) {
      return true
    }
  }

  return false
}

export function isEnumerationType(term: GraphPointer) {
  if (term.has(rdfs.subClassOf, schema.Enumeration).values.length > 0) {
    return true
  }

  const superClasses = term.out(rdfs.subClassOf).toArray()
  for (const superClass of superClasses) {
    if (isEnumerationType(superClass)) {
      return true
    }
  }

  return false
}
