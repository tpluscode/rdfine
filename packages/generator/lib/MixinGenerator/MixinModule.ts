import { Project, SourceFile, VariableDeclarationKind } from 'ts-morph'
import { Context, GeneratedModule } from '../index'
import { SingleContextClownface } from 'clownface'
import { ExternalResourceType, ResourceType, TypeMetaCollection } from '../types'
import { PropertyWriter } from '../property/PropertyWriter'
import { JavascriptProperty } from '../property/JsProperties'
import { getSuperClasses } from './index'

export class MixinModule implements GeneratedModule {
  type: ResourceType
  node: SingleContextClownface;
  superClasses: Array<ResourceType | ExternalResourceType>
  properties: JavascriptProperty[]
  private namespaceImports: Record<string, string> = {}
  private mixinImports: Array<ResourceType | ExternalResourceType> = []

  public constructor(clas: SingleContextClownface, type: ResourceType, superClasses: (ResourceType | ExternalResourceType)[], properties: JavascriptProperty[]) {
    this.node = clas
    this.type = type
    this.superClasses = superClasses
    this.superClasses.forEach(this.addMixinImport.bind(this))
    this.properties = properties
  }

  addMixinImport(type: ExternalResourceType | ResourceType) {
    const mixinImportExists = this.mixinImports.find(current => current.module === type.module)
    if (!mixinImportExists) {
      this.mixinImports.push(type)
    }

    if ('package' in type && !this.namespaceImports[type.package]) {
      this.namespaceImports[type.package] = type.qualifier
    }
  }

  writeModule(project: Project, types: TypeMetaCollection, context: Pick<Context, 'log' | 'prefix' | 'vocabulary' | 'defaultExport'>) {
    context.log.debug(`Generating mixin ${this.type.qualifiedName}`)

    const mixinFile = project.createSourceFile(`${this.type.module}.ts`, {}, { overwrite: true })
    const bundleModule = project.createSourceFile(`bundles/${this.type.module}.ts`, {}, { overwrite: true })
    const bundleIndex = project.getSourceFile('bundles/index.ts') || project.createSourceFile('bundles/index.ts', {}, { overwrite: true })

    const mixinName = this.type.mixinName
    const implName = `${this.type.localName}Impl`
    const interfaceDeclaration = this.createInterface(mixinFile)
    const classDeclaration = this.createMixinFunction(mixinFile, context)

    const propertyWriter = new PropertyWriter({
      interfaceDeclaration,
      classDeclaration,
      context,
      module: this,
    })
    this.properties.forEach(propertyWriter.addProperty.bind(propertyWriter))

    const implementationClass = mixinFile.addClass({
      name: implName,
      extends: `${mixinName}(RdfResourceImpl)`,
    })
    implementationClass.addConstructor({
      parameters: [
        { name: 'arg', type: 'ResourceNode' },
        { name: 'init', type: `Initializer<${this.type.localName}>`, hasQuestionToken: true }],
      statements: [
        'super(arg, init)',
        `this.types.add(${context.prefix}.${this.type.localName})`,
      ],
    })

    const mixinNames = this.superClasses.reduce((mixins, sc) => {
      return [...mixins, sc.type === 'Resource' ? sc.mixinName : sc.qualifiedMixinName]
    }, [this.type.mixinName])
    implementationClass.addProperty({
      isStatic: true,
      isReadonly: true,
      name: '__mixins',
      type: 'Mixin[]',
      initializer: `[${mixinNames.join(', ')}]`,
    })

    mixinFile.addStatements([
      `${mixinName}.appliesTo = ${context.prefix}.${this.type.localName}`,
      `${mixinName}.Class = ${implName}`,
    ])

    this.addImports(mixinFile, context)
    this.generateDependenciesModule(bundleModule, bundleIndex, types)

    return {
      mainModuleExport: this.type.localName,
      mainModuleMixinExport: mixinName,
    }
  }

  private generateDependenciesModule(depsModule: SourceFile, bundleIndex: SourceFile, types: TypeMetaCollection) {
    const exports: string[] = []
    const imported: Map<string, ResourceType | ExternalResourceType> = new Map()
    const name = `${this.type.localName}Bundle`

    bundleIndex.addExportDeclaration({
      moduleSpecifier: `./${this.type.localName}`,
      namedExports: [name],
    })

    depsModule.addImportDeclaration({
      namedImports: ['Mixin'],
      moduleSpecifier: '@tpluscode/rdfine/lib/ResourceFactory',
      isTypeOnly: true,
    })

    const toImport = this.properties.reduce((toImport, prop) => {
      return prop.range
        .reduce((toImport, range) => {
          if (range.type === 'Resource') {
            toImport.push(range)
          }

          return toImport
        }, toImport)
    }, [...this.mixinImports, this.type])

    while (toImport.length) {
      const next = toImport.splice(0, 1)[0]
      if (imported.has(next.mixinName)) continue

      imported.set(next.mixinName, next)
      getSuperClasses(this.node, types)
        .filter(sc => sc.type === 'Resource')
        .forEach(sc => toImport.push(sc))
    }

    imported
      .forEach(mi => {
        if (mi.type !== 'Resource') {
          return
        }

        depsModule.addImportDeclaration({
          namedImports: [`${mi.mixinName}`],
          moduleSpecifier: `.${mi.module}`,
        })
        exports.push(`${mi.mixinName} as Mixin`)
      })

    depsModule.addVariableStatement({
      isExported: true,
      declarationKind: VariableDeclarationKind.Const,
      declarations: [{
        name,
        initializer: `[\n${exports.join(',\n')}]`,
      }],
    })
  }

  private addImports(mixinFile: SourceFile, context: Omit<Context, 'properties'>) {
    const rdfineImports = ['Constructor', 'namespace', 'RdfResource']

    if (Object.keys(this.properties).some(Boolean)) {
      rdfineImports.push('property')
    }

    mixinFile.addImportDeclaration({
      defaultImport: 'RdfResourceImpl',
      namedImports: rdfineImports,
      moduleSpecifier: '@tpluscode/rdfine',
    })

    mixinFile.addImportDeclaration({
      namespaceImport: '$rdf',
      moduleSpecifier: '@rdf-esm/data-model',
    })
    mixinFile.addImportDeclaration({
      namespaceImport: 'RDF',
      moduleSpecifier: 'rdf-js',
      isTypeOnly: true,
    })
    mixinFile.addImportDeclaration({
      namedImports: [context.prefix],
      moduleSpecifier: './lib/namespace',
    })
    mixinFile.addImportDeclaration({
      namedImports: ['Initializer', 'ResourceNode'],
      moduleSpecifier: '@tpluscode/rdfine/RdfResource',
      isTypeOnly: true,
    })
    mixinFile.addImportDeclaration({
      namedImports: ['Mixin'],
      moduleSpecifier: '@tpluscode/rdfine/lib/ResourceFactory',
      isTypeOnly: true,
    })
    mixinFile.addImportDeclaration({
      namespaceImport: context.defaultExport,
      moduleSpecifier: '.',
      isTypeOnly: true,
    })

    Object.entries(this.namespaceImports)
      .forEach(([moduleSpecifier, namespaceImport]) => {
        mixinFile.addImportDeclaration({
          moduleSpecifier,
          namespaceImport,
        })
      })

    this.superClasses
      .filter(sc => sc.type === 'Resource')
      .forEach(superClass => {
        mixinFile.addImportDeclaration({
          moduleSpecifier: superClass.module,
          namedImports: [superClass.mixinName],
        })
      })
  }

  private createMixinFunction(mixinFile: SourceFile, context: Omit<Context, 'properties'>) {
    const mixinFunction = mixinFile.addFunction({
      name: this.type.mixinName,
      typeParameters: [{
        name: 'Base',
        constraint: 'Constructor',
      }],
      parameters: [{
        name: 'Resource',
        type: 'Base',
      }],
      isExported: true,
    })

    const baseClass = this.superClasses
      .reduce((type, superClass) => {
        const mixinName = superClass.type === 'ExternalResource' ? superClass.qualifiedMixinName : superClass.mixinName

        return `${mixinName}(${type})`
      },
      'Resource')

    const className = `${this.type.localName}Class`
    const mixinClass = mixinFunction.addClass({
      name: className,
      extends: baseClass,
      implements: [this.type.localName],
    })

    mixinClass.addDecorator({
      name: 'namespace',
      arguments: [context.prefix],
    })

    mixinFunction.addStatements('return ' + className)

    return mixinClass
  }

  private createInterface(mixinFile: SourceFile) {
    const superInterfaces = this.superClasses
      .map(superClass => `${superClass.qualifiedName}`)

    return mixinFile.addInterface({
      name: this.type.localName,
      isExported: true,
      extends: [...superInterfaces, 'RdfResource'],
    })
  }
}
