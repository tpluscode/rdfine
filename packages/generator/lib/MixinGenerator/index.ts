import { GraphPointer } from 'clownface'
import { Context } from '../index'
import { rdf, rdfs, hydra, owl } from '@tpluscode/rdf-ns-builders'
import { ExternalResourceType, ResourceType, TypeMetaCollection } from '../types'
import { MixinModule } from './MixinModule'
import { findProperties } from '../property'
import { toJavascriptProperties } from '../property/JsProperties'
import { NamedNode } from 'rdf-js'
import TermSet from '@rdfjs/term-set'
import TermMap from '@rdfjs/term-map'

export function getSuperClasses(clas: GraphPointer, types: TypeMetaCollection) {
  return clas.out(rdfs.subClassOf)
    .toArray()
    .reduce<(ResourceType | ExternalResourceType)[]>((selected, candidate) => {
    const found = types.get(candidate)
    if (found && (found.type === 'Resource' || found.type === 'ExternalResource')) {
      selected.push(found)
    }

    return selected
  }, [])
}

function * getProperties(clas: GraphPointer, types: TypeMetaCollection, context: Context, excludedTerms: TermSet) {
  for (const { term, range } of findProperties(clas, context)) {
    if (excludedTerms.has(term.term)) continue

    for (const jsProperty of toJavascriptProperties(term, range, types, context)) {
      yield jsProperty
    }
  }
}

export function findTermsToGenerate(excludedTerms: NamedNode[]) {
  const excluded = new TermSet(excludedTerms)

  return function (types: TypeMetaCollection, context: Context) {
    const terms = new TermMap(context.vocabulary
      .has(rdf.type, [rdfs.Class, hydra.Class, owl.Class])
      .filter(term => types.get(term)?.type === 'Resource')
      .map(pointer => [pointer.term, pointer]))

    return [...terms]
      .map(([, term]) => {
        const meta = types.getOrThrow(term)
        if (meta.type !== 'Resource') {
          throw new Error(`Expected resource type but got ${meta.type}`)
        }

        const props = [...getProperties(term, types, context, excluded)]
        return new MixinModule(
          term,
          meta,
          getSuperClasses(term, types),
          props)
      })
  }
}
