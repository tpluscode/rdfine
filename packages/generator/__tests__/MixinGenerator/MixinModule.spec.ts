import { Project } from 'ts-morph'
import cf, { AnyPointer } from 'clownface'
import $rdf from 'rdf-ext'
import { xsd } from '@tpluscode/rdf-ns-builders'
import { MixinModule } from '../../lib/MixinGenerator/MixinModule'
import { FakeTypeCollection } from '../_helpers/FakeTypeCollection'
import { ex } from '../_helpers/prefix'
import { fakeLog } from '../_helpers/util'

describe('MixinModule', () => {
  let project: Project
  let vocabulary: AnyPointer

  beforeEach(() => {
    project = new Project({
      useInMemoryFileSystem: true,
    })

    vocabulary = cf({ dataset: $rdf.dataset() })
  })

  it('generates a mixin class without props', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates a mixin class with external super classes', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [{
      type: 'ExternalResource' as const,
      mixinName: 'PersonMixin',
      module: '@rdfine/schema/Person',
      qualifiedName: 'Schema.Person',
      package: '@rdfine/schema',
      qualifier: 'Schema',
      qualifiedMixinName: 'Schema.PersonMixin',
    }], [])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates property with automatic type cast when predicate has strict semantics', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'strict' as const,
      termName: 'foo',
      range: [{
        type: 'Resource' as const,
        mixinName: 'FooMixin',
        module: './Foo',
        localName: 'Foo',
        qualifiedName: 'Example.Foo',
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates array property', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource',
      prefixedTerm: 'ex.foo',
      semantics: 'loose',
      termName: 'foo',
      values: ['array'],
      range: [{
        type: 'Resource',
        mixinName: 'FooMixin',
        module: './Foo',
        localName: 'Foo',
        qualifiedName: 'Example.Foo',
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates list property', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource',
      prefixedTerm: 'ex.foo',
      semantics: 'loose',
      termName: 'foo',
      values: ['list'],
      range: [{
        type: 'Resource',
        mixinName: 'FooMixin',
        module: './Foo',
        localName: 'Foo',
        qualifiedName: 'Example.Foo',
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates list/single property', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource',
      prefixedTerm: 'ex.foo',
      semantics: 'loose',
      termName: 'foo',
      values: ['list', 'single'],
      range: [{
        type: 'Resource',
        mixinName: 'FooMixin',
        module: './Foo',
        localName: 'Foo',
        qualifiedName: 'Example.Foo',
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('does not try import self when property has range of enclosing class', () => {
    // given
    const generateType = {
      type: 'Resource' as const,
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }
    const module = new MixinModule(vocabulary.node(ex.Class), generateType, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'strict' as const,
      termName: 'foo',
      range: [generateType],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates property with automatic external type cast when predicate has strict semantics', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'strict' as const,
      termName: 'foo',
      range: [{
        type: 'ExternalResource' as const,
        mixinName: 'FooMixin',
        module: '@rdfine/example/Foo',
        qualifiedName: 'Example.Foo',
        package: '@rdfine/example',
        qualifier: 'Example',
        qualifiedMixinName: 'Example.FooMixin',
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates property with datatype-annotated literal', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'literal' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'strict' as const,
      termName: 'foo',
      range: [{
        type: 'Literal' as const,
        nativeName: 'string',
        nativeType: String,
        datatype: xsd.anyURI,
      }, {
        type: 'Literal' as const,
        nativeName: 'string',
        nativeType: Number,
        datatype: xsd.anyURI,
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates date property', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'literal' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'strict' as const,
      termName: 'foo',
      range: [{
        type: 'Literal' as const,
        nativeName: 'Date',
        nativeType: Date,
        datatype: xsd.date,
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('does not add datatype annotation to literal property when datatype is ambiguous', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'literal' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'strict' as const,
      termName: 'foo',
      range: [{
        type: 'Literal' as const,
        nativeName: 'string',
        nativeType: String,
        datatype: xsd.anyURI,
      }, {
        type: 'Literal' as const,
        nativeName: 'string',
        nativeType: Number,
        datatype: xsd.int,
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('Class.ts')).toMatchSnapshot()
  })

  it('generates a bundle module with property mixin types but skipping external mixins', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [], [{
      term: ex.foo,
      name: 'foo',
      type: 'resource' as const,
      prefixedTerm: 'ex.foo',
      semantics: 'loose' as const,
      termName: 'foo',
      range: [{
        type: 'ExternalResource' as const,
        mixinName: 'FooMixin',
        module: '@rdfine/example/Foo',
        qualifiedName: 'Example.Foo',
        package: '@rdfine/example',
        qualifier: 'Example',
        qualifiedMixinName: 'Example.FooMixin',
      }, {
        type: 'Resource' as const,
        localName: 'Friend',
        module: './Friend',
        qualifiedName: 'Example.Friend',
        mixinName: 'FriendMixin',
      }],
    }])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('bundles/Class.ts')).toMatchSnapshot()
    expect(project.getSourceFile('bundles/index.ts')).toMatchSnapshot()
  })

  it('generates a bundle module with super classes but without externals', () => {
    // given
    const module = new MixinModule(vocabulary.node(ex.Class), {
      type: 'Resource',
      localName: 'Class',
      qualifiedName: 'Example.Class',
      module: './Class',
      mixinName: 'ClassMixin',
    }, [{
      type: 'Resource' as const,
      mixinName: 'SuperMixin',
      qualifiedName: 'Example.Super',
      module: './Example',
      localName: 'Super',
    }, {
      type: 'ExternalResource' as const,
      mixinName: 'ExternMixin',
      qualifiedName: 'Example.Extern',
      module: 'Extern',
      qualifiedMixinName: 'Example.FooExtern',
      package: 'Extern',
      qualifier: 'Foo',
    }], [])

    // when
    module.writeModule(project, new FakeTypeCollection(), {
      prefix: 'ex',
      defaultExport: 'Example',
      log: fakeLog(),
      vocabulary,
    })

    // then
    expect(project.getSourceFile('bundles/Class.ts')).toMatchSnapshot()
    expect(project.getSourceFile('bundles/index.ts')).toMatchSnapshot()
  })
})
