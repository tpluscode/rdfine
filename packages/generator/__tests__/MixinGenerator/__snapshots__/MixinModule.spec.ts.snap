// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MixinModule does not try import self when property has range of enclosing class: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";

export interface Class extends RdfResource {
    foo: Example.Class;
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Resource implements Class {
        @property.resource({ as: [ClassMixin] })
        foo!: Example.Class;
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;

exports[`MixinModule generates a bundle module with property mixin types but skipping external mixins: toMatchSnapshot 1`] = `
"import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import { ClassMixin } from \\"../Class\\";
import { FriendMixin } from \\"../Friend\\";

export const ClassBundle = [
    ClassMixin as Mixin,
    FriendMixin as Mixin];
"
`;

exports[`MixinModule generates a bundle module with property mixin types but skipping external mixins: toMatchSnapshot 2`] = `
"export { ClassBundle } from \\"./Class\\";
"
`;

exports[`MixinModule generates a bundle module with super classes but without externals: toMatchSnapshot 1`] = `
"import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import { SuperMixin } from \\"../Example\\";
import { ClassMixin } from \\"../Class\\";

export const ClassBundle = [
    SuperMixin as Mixin,
    ClassMixin as Mixin];
"
`;

exports[`MixinModule generates a bundle module with super classes but without externals: toMatchSnapshot 2`] = `
"export { ClassBundle } from \\"./Class\\";
"
`;

exports[`MixinModule generates a mixin class with external super classes: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";
import * as Schema from \\"@rdfine/schema\\";

export interface Class extends Schema.Person, RdfResource {
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Schema.PersonMixin(Resource) implements Class {
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin, Schema.PersonMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;

exports[`MixinModule generates a mixin class without props: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";

export interface Class extends RdfResource {
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Resource implements Class {
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;

exports[`MixinModule generates array property: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";

export interface Class extends RdfResource {
    foo: Array<Example.Foo>;
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Resource implements Class {
        @property.resource({ values: 'array' })
        foo!: Array<Example.Foo>;
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;

exports[`MixinModule generates list property: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";

export interface Class extends RdfResource {
    foo: Array<Example.Foo>;
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Resource implements Class {
        @property.resource({ values: 'list' })
        foo!: Array<Example.Foo>;
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;

exports[`MixinModule generates property with automatic external type cast when predicate has strict semantics: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";
import * as Example from \\"@rdfine/example\\";

export interface Class extends RdfResource {
    foo: Example.Foo;
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Resource implements Class {
        @property.resource({ as: [Example.FooMixin] })
        foo!: Example.Foo;
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;

exports[`MixinModule generates property with automatic type cast when predicate has strict semantics: toMatchSnapshot 1`] = `
"import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from \\"@tpluscode/rdfine\\";
import type * as RDF from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";
import type { Initializer, ResourceNode } from \\"@tpluscode/rdfine/RdfResource\\";
import type { Mixin } from \\"@tpluscode/rdfine/lib/ResourceFactory\\";
import type * as Example from \\".\\";

export interface Class extends RdfResource {
    foo: Example.Foo;
}

export function ClassMixin<Base extends Constructor>(Resource: Base) {
    @namespace(ex)
    class ClassClass extends Resource implements Class {
        @property.resource({ implicitTypes: [ex.Foo] })
        foo!: Example.Foo;
    }
    return ClassClass
}

class ClassImpl extends ClassMixin(RdfResourceImpl) {
    constructor(arg: ResourceNode, init?: Initializer<Class>) {
        super(arg, init)
        this.types.add(ex.Class)
    }

    static readonly __mixins: Mixin[] = [ClassMixin];
}
ClassMixin.appliesTo = ex.Class
ClassMixin.Class = ClassImpl
"
`;
